File: tracks/beta/drivers/r/R/scram.R
# ScratchBird-driver
# Copyright (c) 2025-2026 Dalton Calford
#
# Licensed under the Initial Developer's Public License Version 1.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at:
# https://www.firebirdsql.org/en/initial-developer-s-public-license-version-1-0/
sb_scram_client <- function(username) {
  nonce <- openssl::base64_encode(openssl::rand_bytes(18))
  list(
    username = username,
    client_nonce = nonce,
    client_first_bare = "",
    server_signature = NULL
  )
}

raw_xor <- function(a, b) {
  as.raw(bitwXor(as.integer(a), as.integer(b)))
}

hmac_sha256 <- function(key, message) {
  key_raw <- if (is.raw(key)) key else charToRaw(as.character(key))
  msg_raw <- if (is.raw(message)) message else charToRaw(as.character(message))
  block_size <- 64L
  if (length(key_raw) > block_size) {
    key_raw <- openssl::sha256(key_raw)
  }
  if (length(key_raw) < block_size) {
    key_raw <- c(key_raw, as.raw(rep(0L, block_size - length(key_raw))))
  }
  o_key <- raw_xor(key_raw, as.raw(rep(0x5c, block_size)))
  i_key <- raw_xor(key_raw, as.raw(rep(0x36, block_size)))
  inner <- openssl::sha256(c(i_key, msg_raw))
  openssl::sha256(c(o_key, inner))
}

int_to_raw_be <- function(value) {
  con <- rawConnection(raw(), "wb")
  on.exit(close(con))
  writeBin(as.integer(value), con, size = 4, endian = "big")
  rawConnectionValue(con)
}

pbkdf2_hmac_sha256 <- function(password, salt, iterations, keylen) {
  hlen <- 32L
  blocks <- ceiling(keylen / hlen)
  out <- raw()
  for (i in seq_len(blocks)) {
    u <- hmac_sha256(password, c(salt, int_to_raw_be(i)))
    t <- u
    if (iterations > 1) {
      for (j in 2:iterations) {
        u <- hmac_sha256(password, u)
        t <- raw_xor(t, u)
      }
    }
    out <- c(out, t)
  }
  out[seq_len(keylen)]
}

sb_scram_client_first <- function(state) {
  escaped <- gsub("=", "=3D", gsub(",", "=2C", state$username, fixed = TRUE), fixed = TRUE)
  state$client_first_bare <- paste0("n=", escaped, ",r=", state$client_nonce)
  list(state = state, message = paste0("n,,", state$client_first_bare))
}

sb_scram_handle_server_first <- function(state, password, server_first) {
  attrs <- parse_scram_attrs(server_first)
  nonce <- attrs$r
  if (is.null(nonce) || !startsWith(nonce, state$client_nonce)) stop("SCRAM server nonce mismatch")
  salt_b64 <- attrs$s
  iter_str <- attrs$i
  if (is.null(salt_b64) || is.null(iter_str)) stop("SCRAM server-first missing fields")
  iterations <- as.integer(iter_str)
  salt <- openssl::base64_decode(salt_b64)
  salted <- pbkdf2_hmac_sha256(password, salt, iterations, keylen = 32)
  client_key <- hmac_sha256(salted, "Client Key")
  stored_key <- openssl::sha256(client_key)
  client_final_without_proof <- paste0("c=biws,r=", nonce)
  auth_message <- paste(state$client_first_bare, server_first, client_final_without_proof, sep = ",")
  client_signature <- hmac_sha256(stored_key, auth_message)
  client_proof <- raw_xor(client_key, client_signature)
  server_key <- hmac_sha256(salted, "Server Key")
  state$server_signature <- hmac_sha256(server_key, auth_message)
  proof_b64 <- openssl::base64_encode(as.raw(client_proof))
  list(state = state, message = paste0(client_final_without_proof, ",p=", proof_b64))
}

sb_scram_verify_server_final <- function(state, server_final) {
  attrs <- parse_scram_attrs(server_final)
  verifier <- attrs$v
  if (is.null(verifier)) stop("SCRAM server-final missing verifier")
  expected <- openssl::base64_encode(state$server_signature)
  if (verifier != expected) stop("SCRAM server signature mismatch")
  TRUE
}

parse_scram_attrs <- function(message) {
  attrs <- list()
  parts <- strsplit(message, ",", fixed = TRUE)[[1]]
  for (part in parts) {
    if (part == "") next
    kv <- strsplit(part, "=", fixed = TRUE)[[1]]
    if (length(kv) < 2) next
    attrs[[kv[1]]] <- kv[2]
  }
  attrs
}

File: tracks/beta/drivers/dart/lib/src/scram.dart
// ScratchBird-driver
// Copyright (c) 2025-2026 Dalton Calford
//
// Licensed under the Initial Developer's Public License Version 1.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at:
// https://www.firebirdsql.org/en/initial-developer-s-public-license-version-1-0/

import 'dart:convert';
import 'dart:math';
import 'dart:typed_data';

import 'package:crypto/crypto.dart';

import 'errors.dart';

class ScramClient {
  final String username;
  late String clientNonce;
  String? clientFirstBare;
  Uint8List? serverSignature;

  ScramClient(this.username) {
    final rand = Random.secure();
    final bytes = List<int>.generate(18, (_) => rand.nextInt(256));
    clientNonce = base64.encode(bytes);
  }

  String clientFirstMessage() {
    clientFirstBare = 'n=${_escape(username)},r=$clientNonce';
    return 'n,,' + clientFirstBare!;
  }

  String handleServerFirst(String password, String serverFirst) {
    final attrs = _parseAttrs(serverFirst);
    final nonce = attrs['r'] ?? '';
    final saltB64 = attrs['s'] ?? '';
    final iterStr = attrs['i'] ?? '0';
    if (!nonce.startsWith(clientNonce)) {
      throw const ScratchBirdAuthException('SCRAM server nonce mismatch');
    }
    final iterations = int.parse(iterStr);
    final salt = base64.decode(saltB64);
    final salted = _pbkdf2Sha256(utf8.encode(password), salt, iterations, 32);
    final clientKey = _hmacSha256(salted, utf8.encode('Client Key'));
    final storedKey = sha256.convert(clientKey).bytes;
    final clientFinalWithoutProof = 'c=biws,r=$nonce';
    final authMessage =
        '${clientFirstBare!},$serverFirst,$clientFinalWithoutProof';
    final clientSignature = _hmacSha256(storedKey, utf8.encode(authMessage));
    final clientProof = _xor(clientKey, clientSignature);
    final serverKey = _hmacSha256(salted, utf8.encode('Server Key'));
    serverSignature = _hmacSha256(serverKey, utf8.encode(authMessage));
    return '$clientFinalWithoutProof,p=${base64.encode(clientProof)}';
  }

  void verifyServerFinal(String serverFinal) {
    final attrs = _parseAttrs(serverFinal);
    final verifier = attrs['v'] ?? '';
    final expected = base64.encode(serverSignature ?? Uint8List(0));
    if (verifier != expected) {
      throw const ScratchBirdAuthException('SCRAM server signature mismatch');
    }
  }

  String _escape(String input) =>
      input.replaceAll('=', '=3D').replaceAll(',', '=2C');

  Map<String, String> _parseAttrs(String message) {
    final out = <String, String>{};
    for (final part in message.split(',')) {
      final idx = part.indexOf('=');
      if (idx > 0) {
        out[part.substring(0, idx)] = part.substring(idx + 1);
      }
    }
    return out;
  }

  Uint8List _hmacSha256(List<int> key, List<int> data) {
    final hmacSha = Hmac(sha256, key);
    return Uint8List.fromList(hmacSha.convert(data).bytes);
  }

  Uint8List _xor(List<int> left, List<int> right) {
    final out = Uint8List(left.length);
    for (var i = 0; i < left.length; i++) {
      out[i] = left[i] ^ right[i];
    }
    return out;
  }

  Uint8List _pbkdf2Sha256(
      List<int> password, List<int> salt, int iterations, int keyLen) {
    final blocks = (keyLen / 32).ceil();
    final out = BytesBuilder();
    for (var i = 1; i <= blocks; i++) {
      out.add(_pbkdf2F(password, salt, iterations, i));
    }
    final bytes = out.toBytes();
    return Uint8List.sublistView(bytes, 0, keyLen);
  }

  List<int> _pbkdf2F(
      List<int> password, List<int> salt, int iterations, int blockIndex) {
    final block = Uint8List(salt.length + 4);
    block.setAll(0, salt);
    block[block.length - 4] = (blockIndex >> 24) & 0xff;
    block[block.length - 3] = (blockIndex >> 16) & 0xff;
    block[block.length - 2] = (blockIndex >> 8) & 0xff;
    block[block.length - 1] = blockIndex & 0xff;
    var u = _hmacSha256(password, block);
    final out = List<int>.from(u);
    for (var i = 1; i < iterations; i++) {
      u = _hmacSha256(password, u);
      for (var j = 0; j < out.length; j++) {
        out[j] ^= u[j];
      }
    }
    return out;
  }
}
